### PR TITLE
Add tendermint log appender

### LINKED ISSUE #9: feature: tendermint log appender
## Background

`noird` are using two loggers from c++ and go. Eventually `tendermint-go` can be superseded by our own consensus implementation, but it needs to integrate logs into either one at this moment.

## Requirement

`fc::logger` can refer to multiple `fc::appender`s [[1]](https://github.com/noir-protocol/fc/blob/master/src/log/logger.cpp#L22), and iterates all appenders when user prints a log by logger [[2]](https://github.com/noir-protocol/fc/blob/master/src/log/logger.cpp#L66-L68).

It is possible to redirect logs to tendermint logger by implementing `noir::log::tmlog_appender` using tendermint logger [wrappers](https://github.com/noir-protocol/noir/blob/main/src/noir/tendermint/log.cpp).

``` c++
namespace noir::log {

class tmlog_appender : public fc::appender {
public:
  tmlog_appender(const fc::variant& args);
  tmlog_appender();

  void initialize(boost::asio::io_context& io_context) override;
  void log(const fc::log_message& m) override;
};

} // namespace noir::log
```

### PR DESCRIPTION
**Pull request by lazka33**
Resolves #9

### CODE REVIEW

File: src/noir/common/log.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#include <noir/common/log.h>
+
+#include <fc/log/logger_config.hpp>
+
+#include <noir/tendermint/tendermint.h>
```
Comment (conr2d) [inline src/noir/common/log.cpp]:
```suggestion
#include <noir/common/log.h>
#include <noir/tendermint/tendermint.h>
#include <fc/log/logger_config.hpp>
```
Other includes follow the lexicographical sequence of (noir -> 3rd libs -> std libs).

File: src/noir/common/log.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#pragma once
+#include <fc/log/appender.hpp>
+#include <fc/log/log_message.hpp>
+#include <fc/log/logger_config.hpp>
+
+namespace noir::log {
+
+  class tmlog_appender : public fc::appender {
```
Comment (conr2d) [inline src/noir/common/log.h]: What about renaming this and putting under tendermint?

e.g.
``` c++
namespace noir::tendermint::log {

class appender : public fc::appender {
  /* ... */
};

} // namespace noir::tendermint::log
```

(+) we don't indent class declaration.

File: src/noir/common/log.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  public:
+    explicit tmlog_appender( const fc::variant& args );
```
Comment (conr2d) [inline src/noir/common/log.h]: We don't put whitespace around arguments.

```suggestion
    explicit tmlog_appender(const fc::variant& args);
```

File: src/noir/common/log.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    tmlog_appender();
+
+    virtual ~tmlog_appender();
+
+    virtual void initialize( boost::asio::io_service& io_service ) override;
```
Comment (conr2d) [inline src/noir/common/log.h]: - Using virtual and override keywords at the same time is not meaningful. [[the C++ Core Guidelines C.128]](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final)
- `boost::asio::io_service` is deprecated and renamed to `boost::asio::io_context`.

```suggestion
    void initialize(boost::asio::io_context& io_context) override;
```

File: src/noir/common/log.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    virtual void log( const fc::log_message& m ) override;
```
Comment (conr2d) [inline src/noir/common/log.h]: Same here.

```suggestion
    void log(const fc::log_message& m) override;
```

File: src/noir/common/log.cpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        tendermint::log::info(message.c_str());
+        break;
+      case fc::log_level::warn:
+      case fc::log_level::error:
+        tendermint::log::error(message.c_str());
+        break;
+    }
+  }
+
+  void initialize( const char* logger_name ) {
```
Comment (xedonman) [inline src/noir/common/log.cpp]: run formatter here

File: src/noir/common/log.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  };
+
+  static const char* DEFAULT_LOGGER_NAME = "TMLOG";
```
Comment (conr2d) [inline src/noir/common/log.h]: We use snake_case for static variable name too.

File: src/noir/tendermint/tendermint.cpp [hunk 8]
```diff
@@ -1,5 +1,7 @@
 #include <noir/tendermint/tendermint.h>
 
+#include <noir/common/log.h>
```
Comment (conr2d) [inline src/noir/tendermint/tendermint.cpp]:
```suggestion
#include <noir/common/log.h>
#include <noir/tendermint/tendermint.h>
```

File: src/noir/common/log.cpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      );
+
+      fc::logger_config dlc;
+      dlc.name = logger_name;
+      dlc.level = fc::log_level::all;
+      dlc.appenders.emplace_back("default");
+      logging_config.loggers.push_back(dlc);
+
+      return logging_config;
+    }();
```
Comment (conr2d) [inline src/noir/common/log.cpp]: It needs to separate default logger initialization from appender. It could be placed under `noir::common` namespace.

**Review by conr2d — CHANGES_REQUESTED** [review]

File: src/noir/common/log.h (hunk 4 above)
Comment (lazka33) [reply src/noir/common/log.h]: Fixed

File: src/noir/common/log.cpp (hunk 9 above)
Comment (lazka33) [reply src/noir/common/log.cpp]: after offline discussion, no need change

**Review by conr2d — APPROVED** [review]
LGTM.

File: cmd/noird/main.cpp [hunk 10]
```diff
@@ -1,6 +1,7 @@
 #include <noir/commands/commands.h>
 #include <noir/tendermint/tendermint.h>
 #include <appbase/application.hpp>
+#include <noir/common/log.h>
```
Comment (xedonman) [inline cmd/noird/main.cpp]: #include <noir/commands/commands.h>
#include <noir/common/log.h>
#include <noir/tendermint/tendermint.h>
#include <appbase/application.hpp>
seems right order

**Review by xedonman — CHANGES_REQUESTED** [review]
please check comment about include order

**Review by xedonman — APPROVED** [review]
LGTM

**Pull request merged**